addons/membership/models/partner.py

```diff
@@ -0,0 +1,173 @@
+        while len(recursive_ids):
+            self.env.cr.execute('SELECT DISTINCT associate_member_id FROM res_partner WHERE id IN %s', (tuple(recursive_ids),))
+            recursive_ids = filter(None, map(lambda x: x[0], self.env.cr.fetchall()))
+            if not level:
+                raise UserError(_('Error ! You cannot create recursive associated members.'))
+            level -= 1
+        return
+
+    @api.model
+    def _cron_update_membership(self):
+        partners = self.search([('membership_state', '=', 'paid')])
+        if partners:
+            partners._store_set_values(['membership_state'])
+
+    @api.multi
+    def _membership_state(self):
+        """This Function return Membership State For Given Partner. """
+        today_date = date.today()
+        for partner_data in self:
+            partner_data.membership_state = 'none'
+            if partner_data.membership_cancel and today_date > fields.Date.from_string(partner_data.membership_cancel):
+                partner_data.membership_state = 'free' if partner_data.free_member else 'canceled'
+                continue
+            if partner_data.membership_stop and today_date > fields.Date.from_string(partner_data.membership_stop):
+                partner_data.membership_state = 'free' if partner_data.free_member else 'old'
+                continue
+
+            state = 4
+            if partner_data.member_lines_ids:
+                for mline in partner_data.member_lines_ids:
+                    if fields.Date.from_string(mline.date_to) >= today_date:
+                        if mline.account_invoice_line_id and mline.account_invoice_line_id.invoice_id:
+                            mstate = mline.account_invoice_line_id.invoice_id.state
+                            if mstate == 'paid':
+                                state = 0
+                                inv = mline.account_invoice_line_id.invoice_id
+                                for payment in inv.payment_ids:
+                                    if payment.invoice_ids and any(inv.type == 'out_refund' for inv in payment.invoice_ids):
+                                        state = 2
+                                break
+                            elif mstate == 'open' and state != 0:
+                                state = 1
+                            elif mstate == 'cancel' and state != 0 and state != 1:
+                                state = 2
+                            elif (mstate == 'draft' or mstate == 'proforma') and state != 0 and state != 1:
+                                state = 3
+                if state == 4:
+                    for mline in partner_data.member_lines_ids:
+                        if fields.Date.from_string(mline.date_from) < today_date and fields.Date.from_string(mline.date_to) < today_date and mline.date_from <= mline.date_to and (mline.account_invoice_line_id and mline.account_invoice_line_id.invoice_id.state) == 'paid':
```

Comment:
``mline.account_invoice_line_id.invoice_id.state == 'paid'`` is enough for last condition.